core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushdownFilterIntoRowNumber.java

```diff
@@ -0,0 +1,193 @@
+        implements Rule<FilterNode>
+{
+    private static final Capture<RowNumberNode> CHILD = newCapture();
+    private static final Pattern<FilterNode> PATTERN = filter().with(source().matching(rowNumber().capturedAs(CHILD)));
+    private final Metadata metadata;
+    private final DomainTranslator domainTranslator;
+    private final TypeOperators typeOperators;
+
+    public PushdownFilterIntoRowNumber(Metadata metadata, TypeOperators typeOperators)
+    {
+        this.metadata = metadata;
+        this.domainTranslator = new DomainTranslator(metadata);
+        this.typeOperators = typeOperators;
+    }
+
+    @Override
+    public Pattern<FilterNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(FilterNode node, Captures captures, Context context)
+    {
+        Session session = context.getSession();
+        TypeProvider types = context.getSymbolAllocator().getTypes();
+
+        DomainTranslator.ExtractionResult extractionResult = fromPredicate(metadata, typeOperators, session, node.getPredicate(), types);
+        TupleDomain<Symbol> tupleDomain = extractionResult.getTupleDomain();
+
+        RowNumberNode source = captures.get(CHILD);
+        Symbol rowNumberSymbol = source.getRowNumberSymbol();
+        OptionalInt upperBound = extractUpperBound(tupleDomain, rowNumberSymbol);
+
+        if (upperBound.isEmpty()) {
+            return Result.empty();
+        }
+
+        if (upperBound.getAsInt() <= 0) {
+            return Result.ofPlanNode(new ValuesNode(node.getId(), node.getOutputSymbols(), ImmutableList.of()));
+        }
+
+        boolean needRewriteSource = true;
+        if (source.getMaxRowCountPerPartition().isPresent()) {
+            int newRowCountPerPartition = Math.min(source.getMaxRowCountPerPartition().get(), upperBound.getAsInt());
+            // No need to rewrite only source node if rowCountPerPartition results in identical.
+            needRewriteSource = source.getMaxRowCountPerPartition().get() != newRowCountPerPartition;
+            if (needRewriteSource) {
+                source = new RowNumberNode(
+                        source.getId(),
+                        source.getSource(),
+                        source.getPartitionBy(),
+                        source.isOrderSensitive(),
+                        source.getRowNumberSymbol(),
+                        Optional.of(newRowCountPerPartition),
+                        source.getHashSymbol());
+            }
+        }
+        else {
+            // Pushdown the upper bound of the filter
+            source = new RowNumberNode(
+                    source.getId(),
+                    source.getSource(),
+                    source.getPartitionBy(),
+                    source.isOrderSensitive(),
+                    source.getRowNumberSymbol(),
+                    Optional.of(upperBound.getAsInt()),
+                    source.getHashSymbol());
+        }
```

Comment:
I was suggested to remove the method used only once to make the logic clear in the `apply` method in [the previous discussion](https://github.com/trinodb/trino/pull/4242).

Reply:
Based on the past comments, I think the previous person making that comment might have been @martint? @martint is this currently better for you to read? This inlined style with the conditional branching is a lot less readable to me.

Reply:
@erichwang that was me commenting previously, however my suggestion was to return Optional (empty in case maxRowCount isn't updated). Reference comparisons look suspicious to me.

Reply:
@kasiafi, yea, agree that optional works too, and would be more readable than the inlined form here.

Reply:
@kasiafi @erichwang Thank you both! 

I would think we may want to merge this PR if the inlining is not so critical because updating the branch many times on `master` which was drastically changed (e.g. package rename, node rename) was challenging work. The review must be also tough.

If the current PR looks acceptable generally, I'm going to create small follow-up PRs to improve the rule afterward. What do you think about it?

Reply:
I have a slight preference to just get it fixed here since it seems like the last simple change. But if anyone else has any concerns, I don't mind delaying too much.

Reply:
@erichwang @kasiafi I updated the PR accordingly. So sorry for bothering you many times but please take a look again when you get a chance.